component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java

```diff
@@ -2519,4 +2528,133 @@ private long getUnlockTimeInMilliSeconds(AuthenticatedUser authenticatedUser) th
+        eventProperties.put(IdentityEventConstants.EventProperty.OTP_GENERATED_TIME, otpGeneratedTime);
+
+        long expiryTime = otpGeneratedTime + Long.parseLong(getExpireTime(context));
+        eventProperties.put(IdentityEventConstants.EventProperty.OTP_EXPIRY_TIME, expiryTime);
+
+        if (StringUtils.isNotEmpty(request.getHeader(EmailOTPAuthenticatorConstants.X_FORWARDED_FOR))) {
+            eventProperties.put(IdentityEventConstants.EventProperty.CLIENT_IP, request.getHeader(
+                    EmailOTPAuthenticatorConstants.X_FORWARDED_FOR).split(",")[0]);
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug("Getting client-ip from remote address since x-forwarded-for is null.");
+            }
+            eventProperties.put(IdentityEventConstants.EventProperty.CLIENT_IP, request.getRemoteAddr());
+        }
+        Event postOtpGenEvent = new Event(IdentityEventConstants.Event.POST_GENERATE_EMAIL_OTP, eventProperties);
+        try {
+            EmailOTPServiceDataHolder.getInstance().getIdentityEventService().handleEvent(postOtpGenEvent);
+        } catch (IdentityEventException e) {
+            String errorMsg = "An error occurred while triggering post event in Email OTP generation flow. " + e.getMessage();
+            throw new AuthenticationFailedException(errorMsg, e);
+        }
+    }
+
+    /**
+     * Trigger event after validating Email OTP.
+     *
+     * @param request HttpServletRequest.
+     * @param context Authentication context.
+     * @throws AuthenticationFailedException Exception on authentication failure.
+     */
+    private void publishPostEmailOTPValidatedEvent(HttpServletRequest request,
+                                                   AuthenticationContext context) throws AuthenticationFailedException {
+
+        Map<String, Object> eventProperties = new HashMap<>();
+        eventProperties.put(IdentityEventConstants.EventProperty.CORRELATION_ID , context.getCallerSessionKey());
+        AuthenticatedUser authenticatedUser = (AuthenticatedUser) context.getProperty(EmailOTPAuthenticatorConstants
+                .AUTHENTICATED_USER);
+        eventProperties.put(IdentityEventConstants.EventProperty.USER_NAME, authenticatedUser.getUserName());
+        eventProperties.put(IdentityEventConstants.EventProperty.TENANT_DOMAIN, context.getTenantDomain());
+        eventProperties.put(IdentityEventConstants.EventProperty.USER_STORE_DOMAIN, authenticatedUser
+                .getUserStoreDomain());
+        eventProperties.put(IdentityEventConstants.EventProperty.APPLICATION_NAME, context.getServiceProviderName());
+        eventProperties.put(IdentityEventConstants.EventProperty.USER_AGENT, request.getHeader(
+                EmailOTPAuthenticatorConstants.USER_AGENT));
+
+        if (StringUtils.isNotEmpty(request.getHeader(EmailOTPAuthenticatorConstants.X_FORWARDED_FOR))) {
```

Comment:
Use `getClientIpAddress` from `IdentityUtil`

Reply:
Fixed in 7ad7f37f8bbf44021015f9a6b207cdf0219f076a